File: python/householder_experiment.py
#!/usr/bin/env python3
# =============================================================================
#     File: householder_experiment.py
#  Created: 2025-02-12 16:06
#   Author: Bernie Roesler
#
"""
Experiment on p 79 of Davis. Test full Q vs Householder vectors.
"""
# =============================================================================

import matplotlib.pyplot as plt
import numpy as np

from scipy import linalg as la, sparse
from scipy.sparse import linalg as spla

from pathlib import Path

filename = Path('../data/west0479.csv')

dtypes = np.dtype([
    ('rows', np.int32),
    ('cols', np.int32),
    ('vals', np.float64)
])

data = np.genfromtxt(filename, delimiter=',', dtype=dtypes)

A = sparse.coo_matrix(
    (data['vals'],
     (data['rows'] - 1, data['cols'] - 1))
).tocsc()

# Use LU to get the permutation vector
# (inefficient! should have a dedicated colamd interface)
lu = spla.splu(A, permc_spec='COLAMD')
q = lu.perm_c
Aq = A[:, q].toarray()

# Get the QR decomposition
Q, R = la.qr(Aq, mode='full')

# Get the Householder vectors
(Qraw, tau), _ = la.qr(Aq, mode='raw')
V = np.tril(Qraw, -1) + np.eye(Aq.shape[0])
R_ = np.triu(Qraw)

np.testing.assert_allclose(R, R_)

# TODO check that these are correct? Book states "Q 38,070 vs V 3,906"
print("Q.nnz:", sparse.csc_matrix(Q).nnz)
print("V.nnz:", sparse.csc_matrix(V).nnz)

# -----------------------------------------------------------------------------
#         Plots
# -----------------------------------------------------------------------------
fig, axs = plt.subplots(num=1, ncols=3, clear=True)
fig.set_size_inches(12, 4.8, forward=True)
# fig.suptitle('Householder vs Full Q')

ax = axs[0]
ax.spy(A, markersize=1)
ax.set_title('A')

ax = axs[1]
ax.spy(Q, markersize=1)
ax.set_title('Q')

ax = axs[2]
ax.spy(V, markersize=1)
ax.set_title('V')

plt.show()

# =============================================================================
# =============================================================================

File: MATLAB/householder_experiment.m
%===============================================================================
%     File: householder_experiment.m
%  Created: 2025-02-12 16:00
%   Author: Bernie Roesler
%
%  Description: Experiment on p 79 of Davis.
%
%===============================================================================

clear; close all;

COLAMD = false;  % if true, use colamd ordering, else use natural ordering

load west0479
A = west0479;

if COLAMD
    q = colamd(A);
else
    q = 1:size(A, 2);
end

[Q, R] = qr(A(:,q));

[V, beta, p, R2] = cs_qr(A(:, q));
Q2 = cs_qright(V, beta, p, speye(size(V, 1)));

fprintf('nnz(A) = %d\n', nnz(A));
fprintf('nnz(Q) = %d\n', nnz(Q));
fprintf('nnz(R) = %d\n', nnz(R));
fprintf('nnz(V) = %d\n', nnz(V));

% With natural ordering:
% nnz(A) = 1888
% nnz(Q) = 148170
% nnz(R) = 59442
% nnz(V) = 40959
%
% With colamd ordering:
% nnz(A) = 1888
% nnz(Q) = 38764
% nnz(R) = 7599
% nnz(V) = 3909

%-------------------------------------------------------------------------------
%        Plots
%-------------------------------------------------------------------------------
fig = figure(1); hold on
set(fig, 'Position', [300, 200, 900, 640]);

fig_title = 'QR factorization of A';
if COLAMD
    fig_title = [fig_title, ' (colamd ordering)'];
else
    fig_title = [fig_title, ' (natural ordering)'];
end
title(fig_title);

subplot(2, 2, 1); hold on
title('A');
spy(A);

subplot(2, 2, 2); hold on
title('V');
spy(V); 

subplot(2, 2, 3); hold on
title('Q');
spy(Q); 

subplot(2, 2, 4); hold on
title('R');
spy(R); 

%===============================================================================
%===============================================================================
